### PR TITLE
Daniel Deng Week06

### PR DESCRIPTION
**Pull request by sinogermany**

### CODE REVIEW

File: cis194/week6/daniel-deng/src/Exercise04.hs [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+streamRepeat :: a -> Stream a
+streamRepeat = Stream <*> streamRepeat
+
+streamMap :: (a -> b) -> Stream a -> Stream b
+streamMap f (Stream x xs) = Stream (f x) . streamMap f $ xs
+
+streamFromSeed :: (a -> a) -> a -> Stream a
+-- streamFromSeed f seed = Stream seed $ streamFromSeed f (f seed)
+streamFromSeed f = Stream <*> streamFromSeed f . f
```
Comment (sinogermany) [inline cis194/week6/daniel-deng/src/Exercise04.hs]: @shineli1984 which variation is preferred here?

File: cis194/week6/daniel-deng/src/Exercise05.hs [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+import CIS194.Homework06.Exercise04 (streamRepeat, streamMap, streamFromSeed)
+
+nats :: Stream Integer
+nats = streamFromSeed (+1) 0
+
+interleaveStreams :: Stream a -> Stream a -> Stream a
+interleaveStreams (Stream x xs) ys = Stream x . interleaveStreams ys $ xs
+
+ruler :: Stream Integer
+ruler = interleaveStreams (streamRepeat 0) (streamMap (+1) ruler)
```
Comment (sinogermany) [inline cis194/week6/daniel-deng/src/Exercise05.hs]: @shineli1984 I find the original (red line) version reads better as it clearly states `interleaveStreams` is of arity 2. But you've explained the rule of less visual distraction. It applies here too? 

```suggestion
ruler = interleaveStreams (streamRepeat 0) . streamMap (+1) $ ruler
```

File: cis194/week6/daniel-deng/src/Exercise06.hs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+{-# OPTIONS_GHC -fno-warn-missing-methods #-}
+
+module CIS194.Homework06.Exercise06 (fibs3) where
+
+import CIS194.Homework06.Exercise03 (Stream(..), streamToList)
+import CIS194.Homework06.Exercise04 (streamRepeat, streamMap)
+
+import Prelude as P
+
+x :: Stream Integer
```
Comment (sinogermany) [inline cis194/week6/daniel-deng/src/Exercise06.hs]: @shineli1984 is it a standard practice / trick to write a Haskell based DSL in this manner? Or am I thinking too much? 😂

File: cis194/week6/daniel-deng/src/Exercise07.hs [hunk 4]
```diff
@@ -0,0 +1,20 @@
+{-# OPTIONS_GHC -fno-warn-missing-methods #-}
+
+module CIS194.Homework06.Exercise07 (fib4) where
+
+data Mat2x2 a = Mat2x2 a a a a
+
+instance Num a => Num (Mat2x2 a) where
+  Mat2x2 a11 a12 a21 a22 * Mat2x2 b11 b12 b21 b22
```
Comment (sinogermany) [inline cis194/week6/daniel-deng/src/Exercise07.hs]: @shineli1984 I find `Mat2x2` clearer, is it violating the Haskell naming convention by putting a digit in the middle?

File: cis194/week6/daniel-deng/src/Exercise04.hs (hunk 1 above)
Comment (shineli1984) [reply cis194/week6/daniel-deng/src/Exercise04.hs]: The second one. Applicative for the win!

File: cis194/week6/daniel-deng/src/Exercise05.hs (hunk 2 above)
Comment (shineli1984) [reply cis194/week6/daniel-deng/src/Exercise05.hs]: Yes.

File: cis194/week6/daniel-deng/src/Exercise06.hs (hunk 3 above)
Comment (shineli1984) [reply cis194/week6/daniel-deng/src/Exercise06.hs]: I'm not which manner you referring to. Can you give me more details here?

File: cis194/week6/daniel-deng/src/Exercise07.hs (hunk 4 above)
Comment (shineli1984) [reply cis194/week6/daniel-deng/src/Exercise07.hs]: No. It's clearer imo too

**Comment by shineli1984** [comment]
Nice One!

**Pull request merged**